apis/core/v1beta1/openstackcontrolplane_webhook.go

```diff
@@ -192,55 +212,55 @@ func (r *OpenStackControlPlane) ValidateCreateServices(basePath *field.Path) fie
 	errors = append(errors, r.ValidateServiceDependencies(basePath)...)
 
 	// Call internal validation logic for individual service operators
-	if r.Spec.Keystone.Enabled {
+	if r.Spec.Keystone.Enabled && r.Spec.Keystone.Template != nil {
```

Comment:
here and below: Do we rely on the defaulting webhook to make the Template non nil if the Enabled is true? It does not seems right to skip validation (and therefore accept) the request if Enabled = True and Template = nil.

// after looking at the defaulting webhook changes

As far as I see the defaulting webhook always ensures the template is not nil. (Note that defaulting webhook runs before the validating hook https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks). So instead of ignoring the validation if the template is nil we should return an error if the template is nil as it never should be.

Also I think we should validate the Template even if the service is not enabled now as we make sure that the defaulting is correct for the non enabled services too.

Reply:
correct, the defaulting runs before the validation, so we could remove it here for validate create. but the initialized template is not written to etcd as `template: {}`. As mentioned in the note at the bottom of the description, to properly initialize and have defaulting you'd still need to pass in `template: {}` for the service. but for disabled services, especially for the service which are per default disabled, or set as disabled I'd like to not have the details set as they are not necessary.

Reply:
let me rephrase what I am trying to achieve. If someone disables the service (or is disabled per default). the resulting service section should not include a defaulted template section:

```
  keystone:
    enabled: false
```

if one enables the service, or provides a template in a disabled service we should defaulted. now I think the validation could happen when the service gets enabled?

Reply:
oh seems there is an issue with the what I planned to do ... if one deploys with disabled and we have no template, then performs an update, we'd not properly validate the update. we'd have to initialize old instead of skip it.

Reply:
> let me rephrase what I am trying to achieve. If someone disables the service (or is disabled per default). the resulting service section should not include a defaulted template section:
> 
> ```
>   keystone:
>     enabled: false
> ```

OK I see now. I think I misunderstood the condition:
```golang
	if r.Spec.Keystone.Enabled || r.Spec.Keystone.Template != nil {
		if r.Spec.Keystone.Template == nil {
			r.Spec.Keystone.Template = &keystonev1.KeystoneAPISpecCore{}
		}
		r.Spec.Keystone.Template.Default()
	}
```
So if the keystone is disabled and the template is not provided then the defaulting is skipped. 

I won't ask to change this to the following as it is complicated in a different way. 
```golang
	if r.Spec.Keystone.Enabled && r.Spec.Keystone.Template == nil {
		r.Spec.Keystone.Template = &keystonev1.KeystoneAPISpecCore{}
        }
        if r.Spec.Keystone.Template != nil {
		r.Spec.Keystone.Template.Default()
        }
```

So this is good as is.